unit_tests/test_aggregator.py

```diff
@@ -0,0 +1,415 @@
+        aggregator_explicit.websocket_thread.start.assert_called_once()
+
+    @staticmethod
+    def test_start_websocket_connection__callback_thread(aggregator_explicit):
+        aggregator_explicit.start_websocket_connection()
+        assert aggregator_explicit.callback_thread == "test_set_callback_thread"
+
+
+class TestGetUuidFromAreaName:
+    """Includes methods which test get_uuid_from_area_name method."""
+
+    @staticmethod
+    def test_get_uuid_from_area_name__not_empty_dict(aggregator_explicit):
+        aggregator_explicit.area_name_uuid_mapping = {"TestArea": ["test_area_uuid"]}
+        area_name = "TestArea"
+
+        patch("gsy_e_sdk.aggregator.get_uuid_from_area_name_in_tree_dict",
+              return_value="test_area_uuid")
+        assert aggregator_explicit.get_uuid_from_area_name(area_name) == "test_area_uuid"
+
+    @staticmethod
+    def test_get_uuid_from_area_name__empty_dict(aggregator_explicit):
+        aggregator_explicit.area_name_uuid_mapping = {}
+        area_name = "TestArea"
+
+        patch("gsy_e_sdk.aggregator.get_uuid_from_area_name_in_tree_dict",
+              return_value="test_area_uuid")
+        assert aggregator_explicit.get_uuid_from_area_name(area_name) is None
+
+
+class TestCalculateGridFee:
+    """Includes methods which test calculate_grid_fee method."""
+
+    @staticmethod
+    def test_calculate_grid_fee__called(aggregator_explicit):
+        with patch("gsy_e_sdk.aggregator.GridFeeCalculation.calculate_grid_fee") as mocked_method:
+            aggregator_explicit.calculate_grid_fee("start_market",
+                                                   "target_market",
+                                                   "current_market_fee")
+            mocked_method.assert_called_with("start_market",
+                                             "target_market",
+                                             "current_market_fee")
+
+    @staticmethod
+    def test_calculate_grid_fee__return(aggregator_explicit):
+        with patch("gsy_e_sdk.aggregator.GridFeeCalculation.calculate_grid_fee",
+                   return_value="test_ok"):
+            ret_val = aggregator_explicit.calculate_grid_fee("start_market",
+                                                             "target_market",
+                                                             "current_market_fee")
+            assert ret_val == "test_ok"
+
+
+class TestExecuteBatchCommands:
+    """Includes methods which test execute_batch_commands method."""
+
+    @staticmethod
+    def test_execute_batch_commands__command_buffer_length_zero(aggregator_explicit):
+        with patch("gsy_e_sdk.aggregator.ClientCommandBuffer.buffer_length",
+                   new_callable=PropertyMock, return_value=0):
+            assert aggregator_explicit.execute_batch_commands() is None
+
+    @staticmethod
+    def test_execute_batch_commands__execute_batch_called(aggregator_explicit):
```

Comment:
`execute_batch_command` is an outgoing query, so we don't need to test it. Testing it would bind us to the implementation. In this case, if `execute_batch` is not called, our tests will fail in any case, because they depend on its return value. So I think that this test is not necessary.

If you want to know a bit more about the context of the "outgoing query" that I mentioned, you can check this great presentation by Sandi Metz: [The Magic Tricks of Testing](https://www.youtube.com/watch?v=URSWYvyc42M). You can also find a recap in our workshops documentation.

Reply:
> In this particular case, IMO part of the function's logic is this execute_batch method being called and this is what is being tested here.

All lines of a function are part of that function's logic. So in theory we should test each line. However, the presentation describes scenarios in which not all lines should be tested. 
You're right when you say that `self._client_command_buffer.execute_batch()` is being called. However, do we need to test this? This line is defining an outgoing message, so there are two cases covered by the presentation:
1. outgoing command
2. outgoing query

**Outgoing commands** cause side-effects on other components of the architecture (for example, it might send a message in a websocket). If they are not called, something will break somewhere and our unit test will have no way to know about it (because we're not testing the distant side-effects in unit tests). For this reason, we want to make sure that our unit under test is calling this method that causes the side-effect: we therefore set an expectation using a mock (assert_called or whatever).
**Outgoing queries**, instead, do not cause side-effects. They return some value, that's it. We need that value because we will use it to do something that depends on it. So if that value is wrong, the method we're testing will return the wrong output (and it will fail, if our test is good). That's why we don't need to set expectations using mocks: if the returned value is wrong, our test will fail. And we won't have created a strict binding with the implementation of the other class.
Our case is the latter IMO. `self._client_command_buffer.execute_batch()` is an outgoing query command. `execute_batch()` doesn't trigger side effects (BTW, the name of this method is confusing because it's not really "executing" anything). So I think we don't need to make assertions about it. If it will return the wrong value, our test should fail.

There are exceptions to this. For example, if `.execute_batch()` was running a very time-consuming computation, we should avoid running that computation in each test. In these cases we use stubs (in python I suppose they're very similar to mocks).

Reply:
Ok, that makes sense. 
However I'm still a little confused about one thing:

Let's say we're testing a method which calls an outgoing query. If we were to let the test/function execute the query (because it might not consume that much time or resources) and use the return value to make any assertion about the following logic of the method being tested, would not this test be bind to the correct behavior of the outgoing query? Therefore wouldn't we be indirectly testing as well the outgoing query itself?
Wouldn't this avoid the test for this method from being a **unit** test?

Regarding your side note, this exception would also apply to external connections, wouldn't it?

Reply:
Extending a bit on my first comment: We should at least mock the return value of this outgoing query, shouldn't we?(just in order to not bind the test to it)
Or would this also somehow _bind us to the implementation_?

Reply:
I've been asking myself the same question many times, and the only answer I could find was: it depends. Tests should be flexible, so that if we change the implementation of some related class they shouldn't fail. That's why we don't make assertions on values returned by outgoing queries: they're redundant and bind us to the implementation. 
Mocking the return value (or creating a "stub", to be precise) would also bind us to the implementation somehow, but I think that in some cases this is unavoidable. IMO, this is the case of long computations, external queries etc.

Reply:
I understand. Tests should then be written as flexible as possible, but that does not mean the can be completely DRY all times. 
For example, in this test suite I faced a case in which the outcome of an outgoing query determined the logic flow in the method being tested. In this case I patched the outgoing query output just to test that both logic paths would be working as expected with each of the possible return values. 
Would you say there is a way around this? Or is this one of the cases you mention here?

> Mocking the return value (or creating a "stub", to be precise) would also bind us to the implementation somehow, but I think that in some cases this is unavoidable.